tests/Unit/DataStructures/VectorImplTestHelper.hpp

```diff
@@ -366,5 +376,338 @@ void vector_test_math_after_move(
+}
+
+// enum for clarity in integral_constant below for choosing between execution
+// paths when assembling the argument list for the math testing utility
+namespace TestFunctionsWithVectorArgumentsChoices {
+struct FirstInplace {};
+struct Continuing {};
+struct Done {};
+}  // namespace TestFunctionsWithVectorArgumentsChoices
+
+// helper struct implementation for choosing the next branch for assembling the
+// argument list for the math testing utility. This is done to use pattern
+// matching instead of SFINAE.
+template <bool IsDone, bool IsFirstAssignment>
+struct next_choice_for_test_functions_recursion_impl;
+
+// if the size of the vector list is the same as the size of the distribution
+// bound list, then assembly is done
+template <bool IsFirstAssignment>
+struct next_choice_for_test_functions_recursion_impl<true, IsFirstAssignment> {
+  using type = TestFunctionsWithVectorArgumentsChoices::Done;
+};
+
+// if we are assembling the first argument and the test type is Inplace, then
+// treat it separately
+template <>
+struct next_choice_for_test_functions_recursion_impl<false, true> {
+  using type = TestFunctionsWithVectorArgumentsChoices::FirstInplace;
+};
+
+// otherwise, continue assembling as ordinary
+template <>
+struct next_choice_for_test_functions_recursion_impl<false, false> {
+  using type = TestFunctionsWithVectorArgumentsChoices::Continuing;
+};
+
+// helper function for easily determining the next branch of the call operator
+// for TestFunctionsWithVectorArgumentsImpl
+template <typename VectorList, typename BoundList, TestKind Test>
+using next_choice_for_test_functions_recursion =
+    typename next_choice_for_test_functions_recursion_impl<
+        tmpl::size<VectorList>::value == tmpl::size<BoundList>::value,
+        tmpl::size<VectorList>::value == 0 and Test == TestKind::Inplace>::type;
+
+// dispatch function to recursively assemble the arguments and wrappers for
+// calling the operator tester `test_function_on_vector_operands`.
+//
+// `UniqueTypeList` is a tmpl::list which stores the set of unique types to
+// test the functions with.
+//
+// `FirstOperand` is the first operand type, used when the Test is
+// `TestKind::Inplace`, as inplace operators have specific demands on the
+// first operand.
+template <TestKind Test, typename UniqueTypeList, typename FirstOperand>
+struct TestFunctionsWithVectorArgumentsImpl {
+  // base case: the correct number of operand types has been obtained in
+  // `Operands`, so this calls the test function with the function, bounds,
+  // reference wrap, and operand type information.
+  template <typename Function, typename... DistBounds, typename... Wraps,
+            typename... Operands>
+  void operator()(
+      const Function& function, const std::tuple<DistBounds...>& bounds,
+      const std::tuple<Wraps...>& wraps,
+      const std::tuple<Operands...>&& operands,
+      TestFunctionsWithVectorArgumentsChoices::Done /*done*/) noexcept {
```

Comment:
The `/*done*/` etc. commented out arg names we've typically been named `/*meta*/` to indicate it's used for metaprogramming only :)